File: technical-tests/backend/milan-jovanovic/source-code-async-locking/typescript-before/currency-conversion.ts
import { CurrencyApiClient } from "./currency-api-client";
import { ExchangeRateResponse } from "./exchange-rate-response";


class CurrencyConversion {
  async handle(currencyCode: string, amount: number, currencyClient: CurrencyApiClient): Promise<ExchangeRateResponse> {
    if (!currencyCode || currencyCode.length !== 3 || !currencyCode.toUpperCase().match(/^[A-Z]{3}$/)) {
      throw new Error("Currency code must be a 3-letter uppercase code (e.g., EUR, GBP)");
    }
    if (amount <= 0) {
      throw new Error("Amount must be a positive number");
    }

    const rate = await currencyClient.getExchangeRate(currencyCode);
    if (!rate) {
      throw new Error(`Exchange rate for ${currencyCode} not found or API error occurred`);
    }

    const convertedAmount = amount * rate;
    return {
      currency: currencyCode,
      baseCurrency: "USD",
      rate: rate,
      amount: amount,
      convertedAmount: convertedAmount,
    };
  }
}

File: technical-tests/backend/milan-jovanovic/source-code-async-locking/python-before/currency_conversion.py
from decimal import decimal
from currency_api_client import CurrencyApiClient
from exchange_rate_response import ExchangeRateResponse


class CurrencyConversion:

    @staticmethod
    async def handle(
        currency_code: str, amount: decimal, currency_client: CurrencyApiClient
    ) -> ExchangeRateResponse:
        if not currency_code or len(currency_code) != 3 or not currency_code.isupper():
            raise ValueError(
                "Currency code must be a 3-letter uppercase code (e.g., EUR, GBP)"
            )
        if amount < 0:
            raise ValueError("Amount must be a positive number")
        rate = await currency_client.get_exchange_rate(currency_code)
        if rate is None:
            raise ValueError(f"Exchange rate for {currency_code} not found")
        
        converted_amount = amount * rate
        return ExchangeRateResponse(
            currency=currency_code,
            base_currency="USD",
            rate=rate,
            amount=amount,
            converted_amount=converted_amount,
        )
